src/util/evented.js

```diff
@@ -124,6 +124,12 @@ export class Evented {
 
             // make sure adding or removing listeners inside other listeners won't cause an infinite loop
             const listeners = this._listeners && this._listeners[type] ? this._listeners[type].slice() : [];
+
+            // Check if onClose is an event listener and move to front of list to fire first
+            if (listeners.length > 1 && listeners[1].name === "bound _onClose") {
```

Comment:
It may work for this specific use case, but this solution is brittle, with possible cases where it breaks and it's very hard to debug — for example, what if a user puts multiple `click` listeners on a popup? Then `_onClose` will not be the second, but later in the array. What if a user adds a listener with the same name? What if a different class in GL JS (not popup) uses the same method name (`_onClose`) and the execution order is swapped unexpectedly?

I think we should find a solution that's more explicit and can't produce unexpected side effects — e.g. firing a `preclick` event before `click` [here](https://github.com/mapbox/mapbox-gl-js/blob/7d2f272e4cf478c286fb115cc7550a19ea477f91/src/ui/handler/map_event.js#L43) and then subscribing to it for `Popup` `closeOnClick`. This is how Leaflet handles it as seen [here](https://github.com/Leaflet/Leaflet/blob/master/src/map/Map.js#L1388-L1397).

Reply:
Thanks so much for the feedback! Yes I wasn't confident on this initial approach. I appreciate sending over the resources and example from leaflet. I implemented a preclick that works manually with the debug file but is breaking some of the unit tests in test/unit/ui/map_events.test.js that I'm looking into now.